### PR TITLE
Add OSS::PowerSelect

### PR DESCRIPTION
**Pull request by JulienVannier66**
### What does this PR do?

Add an `OSS::PowerSelect` and tests
Add documentation for new `inline` arg of `OSS::InfiniteSelect`
Add `preventDefault` for `UpfCheckbox`

Related to: https://github.com/upfluence/backlog/issues/1107

### What are the observable changes?
New component

### Good PR checklist

- [x] Title makes sense
- [x] Is against the correct branch
- [x] Only addresses one issue
- [x] Properly assigned
- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Migrated touched components to Glimmer Components
- [x] Properly labeled

### CODE REVIEW

File: addon/components/o-s-s/power-select.hbs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<div class="upf-power-select fx-col" ...attributes {{did-insert (fn this.checkBlockPresence (has-block 'selected-item') 'selected-item')}}
+  {{did-insert (fn this.checkBlockPresence (has-block 'option-item') 'option-item')}}>
```
Comment (phndiaye) [inline addon/components/o-s-s/power-select.hbs:2]: can you invert the params of `checkBlockPresence`, it makes more sense to have the name first, then the value.
Also, what about naming it `ensureBlockPresence` given that it throws an error?

File: addon/components/o-s-s/power-select.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+export default class OSSPowerSelect extends Component<OSSPowerSelectArgs> {
+  @tracked hasScrollbar = false;
+
+  resizeObserver: ResizeObserver;
+
+  constructor(owner: any, args: OSSPowerSelectArgs) {
+    super(owner, args);
+
+    this.resizeObserver = new ResizeObserver((entries: ResizeObserverEntry[]) => {
+      let observedArrayInput = entries.find((entry) => {
```
Comment (phndiaye) [inline addon/components/o-s-s/power-select.ts:29]: const

**Comment by phndiaye** [comment]
LGTM otherwise

File: addon/components/o-s-s/power-select.hbs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<div class="upf-power-select fx-col" ...attributes {{did-insert (fn this.ensureBlockPresence 'selected-item' (has-block 'selected-item'))}}
+  {{did-insert (fn this.ensureBlockPresence 'option-item' (has-block 'option-item'))}}>
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.hbs:2]: Can't you send both "has-block booleans check" at the same time ? 🤔 
`{{did-insert (fn this.ensureBlockPresence 'selected-item' (has-block 'selected-item') (has-block 'option-item'))}}`
And have the `ensureBlockPresence` method take two params ?

File: addon/components/o-s-s/power-select.hbs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  <div class="upf-power-select__array-container margin-bottom-xxx-sm">
+    <div class="array-input-container {{if @selectedItems 'fx-row fx-xalign-center' 'fx-col fx-malign-center'}}
+                {{if this.hasScrollbar 'has-overflow'}}" {{did-insert this.observeResize}} {{will-destroy this.unobserveResize}}>
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.hbs:5]: Nit: alignment is wrong

File: addon/components/o-s-s/power-select.hbs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      {{#each @selectedItems as |selectedItem index|}}
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.hbs:6]: `index` isn't used.

File: addon/components/o-s-s/power-select.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      }
+    });
+  }
+
+  get placeholder(): string {
+    return this.args.placeholder ?? '';
+  }
+
+  @action
+  observeResize(element: Element) {
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.ts:44]: `observeResize(element: Element): void {`

File: addon/components/o-s-s/power-select.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    this.resizeObserver.observe(element);
+  }
+
+  @action
+  unobserveResize(element: Element) {
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.ts:49]: `unobserveResize(element: Element): void {`

File: addon/components/o-s-s/power-select.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @action
+  removeItem(item: any) {
+    this.args.onChange(item, 'deletion');
+  }
+
+  @action
+  onSelect(item: any) {
+    this.args.onChange(item, 'selection');
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.ts:67]: voids

File: app/styles/power-select.less [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  .upf-power-select__array-container {
+    position: relative;
+    height: 100%;
+    .border;
+    .border-color-default;
+    .border-radius-default;
+
+    .infinite-shadow-scroll {
+      position: absolute;
+      content: '';
```
Comment (Miexil) [inline app/styles/power-select.less:13]: `content` set to empty is needed for the effect to show right ?

File: tests/integration/components/o-s-s/power-select-test.ts [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+      });
+
+      await render(hbs`
+        <OSS::PowerSelect @onSearch={{this.onSearch}}>
+          <:selected-item as |selectedItem|>
+            {{selectedItem.name}}
+          </:selected-item>
+        </OSS::PowerSelect>
+      `);
+    });
```
Comment (Miexil) [inline tests/integration/components/o-s-s/power-select-test.ts:41]: Move the first two tests to a 'It doesn't render' or 'It throws an error' module

File: tests/integration/components/o-s-s/power-select-test.ts [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+  });
+
+  module('with @onChange', (hooks) => {
+    hooks.beforeEach(function () {
+      this.selectedItems = ['value1', 'value2'];
+      this.items = ['value1', 'value2'];
+      this.onChange = onChange;
+    });
+
+    test('delete selected item call onChange with delete operation', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/power-select-test.ts:147]: `calls onChange` or `triggers onChange`

File: tests/integration/components/o-s-s/power-select-test.ts [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+          </:option-item>
+        </OSS::PowerSelect>
+      `);
+
+      await click('.input-array-tag:nth-child(2) i');
+
+      assert.ok(this.onChange.calledWith('value2', 'deletion'));
+    });
+
+    test('select item call onChange with selection operation', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/power-select-test.ts:165]: `selecting item calls ...`

File: tests/integration/components/o-s-s/power-select-test.ts [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+          </:option-item>
+        </OSS::PowerSelect>
+      `);
+
+      assert.dom('.upf-infinite-select input').hasAttribute('placeholder', 'searchPlaceholder');
+    });
+  });
+
+  module('with @onSearch', () => {
+    test('search is correctly call', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/power-select-test.ts:203]: `Search is correctly called`

**Review by phndiaye — APPROVED** [review]

File: addon/components/o-s-s/power-select.hbs (hunk 3 above)
Comment (JulienVannier66) [reply addon/components/o-s-s/power-select.hbs:2]: Yes I can try

File: app/styles/power-select.less (hunk 9 above)
Comment (JulienVannier66) [reply app/styles/power-select.less:13]: Yes!

File: addon/components/o-s-s/power-select.ts [hunk 14]
```diff
@@ -41,29 +41,32 @@ export default class OSSPowerSelect extends Component<OSSPowerSelectArgs> {
...
-  unobserveResize(element: Element) {
+  unobserveResize(element: Element): void {
     this.resizeObserver.unobserve(element);
   }
 
   @action
-  ensureBlockPresence(name: string, hasNamedBock: boolean) {
-    if (!hasNamedBock) {
-      throw new Error(`[component][OSS::PowerSelect] You must pass ${name} named block`);
+  ensureBlockPresence(hasSelectedItem: boolean, hasOptionItem: boolean): void | never {
```
Comment (Miexil) [inline addon/components/o-s-s/power-select.ts:54]: I hadn't raised this one as apparently checking against `throws` had not been introduced into typescript until recently.
As long as we are not getting compilation issues on the CI leave it this way. If it fails though, go ahead and remove the method type here 😉

**Review by Miexil — APPROVED** [review]
LGTM

**Pull request merged**